packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -0,0 +1,260 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+
+class FocusNode extends ChangeNotifier {
+  FocusScopeNode _parent;
+  FocusManager get _manager => _parent?._manager;
```

Comment:
how deep are your trees? I guess there'll be 3 or 4 deep, no more, so this is probably manageable.
If they get longer than this we might want to consider using the attach/detach pattern.